lightning/mydump/router.go

```diff
@@ -0,0 +1,365 @@
+		// source file pattern, matches files like '{schema}.{table}.0001.{sql|csv}'
+		{`(?i)^(?:[^/]*/)*([^/.]+)\.(.*?)(\.[0-9]+)?\.(sql|csv)$`, "$1", "$2", "$4", "$3", ""},
+	}
+)
+
+// // RouteRule is a rule to route file path to target schema/table
+type FileRouter interface {
+	// Route apply rule to path. Return nil if path doesn't math route rule
+	Route(path string) *RouteResult
+}
+
+// chainRouters aggregates multi `FileRouter` as a router
+type chainRouters []FileRouter
+
+func (c chainRouters) Route(path string) *RouteResult {
+	for _, r := range c {
+		res := r.Route(path)
+		if res != nil {
+			return res
+		}
+	}
+	return nil
+}
+
+func Parse(cfg []*config.FileRouteRule) (FileRouter, error) {
+	res := make([]FileRouter, 0, len(cfg))
+	p := &regexRuleParser{}
+	for _, c := range cfg {
+		p.rule = &RegexRouteRule{}
+		if err := p.Parse(c); err != nil {
+			return nil, err
+		}
+		res = append(res, p.rule)
+	}
+	return chainRouters(res), nil
+}
+
+func DefaultRouter() FileRouter {
+	router, _ := Parse(defaultFileRouteRules)
+	return router
+}
+
+type RegexRouteRule struct {
+	pattern    *regexp.Regexp
+	extractors []regexExtractor
+}
+
+func (r *RegexRouteRule) Route(path string) *RouteResult {
+	if !r.pattern.MatchString(path) {
+		return nil
+	}
+
+	indexes := r.pattern.FindStringSubmatchIndex(path)
+	result := &RouteResult{}
+	for _, e := range r.extractors {
+		if !e.Extract(r.pattern, path, indexes, result) {
+			return nil
+		}
+	}
+	return result
+}
+
+type regexRuleParser struct {
+	rule *RegexRouteRule
+	r    *config.FileRouteRule
+}
+
+func (p *regexRuleParser) Parse(r *config.FileRouteRule) error {
+	p.r = r
+	pattern, err := regexp.Compile(p.r.Pattern)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	p.rule.pattern = pattern
+
+	err = p.parseFieldExtractor("type", p.r.Type, func(result *RouteResult, value string) {
+		result.Type = parseSourceType(value)
+	})
+	if err != nil {
+		return err
+	}
+	// ignore pattern needn't parse other fields
+	if p.r.Type == TypeIgnore {
+		return nil
+	}
+
+	err = p.parseFieldExtractor("schema", p.r.Schema, func(result *RouteResult, value string) {
+		result.Schema = value
+	})
+	if err != nil {
+		return err
+	}
+
+	// special case: when the pattern is for db schema, should not parse table name
+	if p.r.Type != SchemaSchema {
+		err = p.parseFieldExtractor("table", p.r.Table, func(result *RouteResult, value string) {
+			result.Name = value
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(p.r.Key) > 0 {
+		err = p.parseFieldExtractor("key", p.r.Key, func(result *RouteResult, value string) {
+			result.Key = value
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(p.r.Compression) > 0 {
+		err = p.parseFieldExtractor("compression", p.r.Compression, func(result *RouteResult, value string) {
+			result.Compression = parseCompressionType(value)
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *regexRuleParser) parseFieldExtractor(
```

Comment:
why do we need this function?

Reply:
This method is use to parse each field in `RegexRouteRule `

Reply:
@glorv  i mean why do we need to perform such parsing? we could just make every field a `patExpander`. the current checks make the code very long 🤔

Reply:
e, since the regex is kind of complex, I think we'd better do some check carefully to avoid filter out source files due to the wrong configed patterns. I will update the check to cover these patterns